crates/kornia-imgproc/src/flip.rs

```diff
@@ -27,43 +33,60 @@ use rayon::{iter::ParallelIterator, slice::ParallelSliceMut};
 /// )
 /// .unwrap();
 ///
-/// let flipped: Image<f32, 3> = horizontal_flip(&image).unwrap();
+/// let mut flipped = Image::<f32, 3>::from_size_val(
+///     ImageSize {
+///         width: 2,
+///         height: 3,
+///     },
+///     0.0
+/// )
+/// .unwrap();
 ///
-/// assert_eq!(flipped.size().width, 2);
-/// assert_eq!(flipped.size().height, 3);
+/// horizontal_flip(&image, &mut flipped).unwrap();
 /// ```
-pub fn horizontal_flip<T, const C: usize>(src: &Image<T, C>) -> Result<Image<T, C>, ImageError>
+pub fn horizontal_flip<T, const C: usize>(
+    src: &Image<T, C>,
+    dst: &mut Image<T, C>,
+) -> Result<(), ImageError>
 where
     T: SafeTensorType,
 {
-    let mut dst = src.clone();
+    if src.size() != dst.size() {
+        return Err(ImageError::InvalidImageSize(
+            src.cols(),
+            src.rows(),
+            dst.cols(),
+            dst.rows(),
+        ));
+    }
 
     dst.as_slice_mut()
         .par_chunks_exact_mut(src.cols() * C)
-        .for_each(|row| {
-            let mut i = 0;
-            let mut j = src.cols() - 1;
-            while i < j {
+        .zip_eq(src.as_slice().par_chunks_exact(src.cols() * C))
+        .for_each(|(dst_row, src_row)| {
+            let n = src.cols() - 1;
+            for i in 0..=n {
```

Comment:
the previous two pointer approach i believe it should be fast as you will cut by half the num of iterations here

Reply:
and one more improvement is to make the swap of the pixel [T; C] at once using `copy_from_slice` or similar in order to remove the inner 0..C loop

Reply:
> the previous two pointer approach i believe it should be fast as you will cut by half the num of iterations here

The reason I had to change the loop (full iteration and adding src) was because I removed `let mut dst = src.clone();` so that we can take in the dst as a parameter.  The other alternative is to keep the loop as as, but initially copy from src to dst. Is this what you prefer?

> and one more improvement is to make the swap of the pixel [T; C] at once using `copy_from_slice` or similar in order to remove the inner 0..C loop

This is what I could come up with:

    dst.as_slice_mut()
        .par_chunks_exact_mut(src.cols() * C)
        .zip_eq(src.as_slice().par_chunks_exact(src.cols() * C))
        .for_each(|(dst_row, src_row)| {
            dst_row
                .chunks_exact_mut(C)
                .zip(src_row.chunks_exact(C).rev())
                .for_each(|(dst_pixel, src_pixel)| {
                    dst_pixel.copy_from_slice(src_pixel);
                })
        });
If it is the above, then the change for `vertical_flip` is just to put `.rev()` in the first zip instead of the second.

--
Am I missing your point or going in the wrong direction? Please let me know which you prefer regarding both points.

Reply:
This might be valid too. An alternative could be what you had but use `n//2` instead, then
```rust
let (idx_i, idx_j) = ( i * C + c, (n-1) * C + c);
dst_row[idx_i] = src_row[idx_i];
dst_row[idx_j] = src_row[idx_j];
```
and somehow avoid the C inner loop using the built in slice copy.

Last, to make sure what’s the best, extend the benchmarks and compare the implementations. Would be an interesting exercise to understand how rust behaves in this case iterators vs index access. Maybe add a `bench_flip.rs` in https://github.com/kornia/kornia-rs/tree/main/crates/kornia-imgproc/benches

Reply:
Thanks, I hadn't thought of using the half loop like that.
I'll try benchmarking.